hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/BlobRenameHandler.java

```diff
@@ -0,0 +1,452 @@
+import static org.apache.hadoop.fs.azurebfs.contracts.services.AzureServiceErrorCode.COPY_BLOB_ABORTED;
+import static org.apache.hadoop.fs.azurebfs.contracts.services.AzureServiceErrorCode.COPY_BLOB_FAILED;
+import static org.apache.hadoop.fs.azurebfs.contracts.services.AzureServiceErrorCode.RENAME_DESTINATION_PARENT_PATH_NOT_FOUND;
+
+/**
+ * Orchestrator for rename over Blob endpoint. Handles both directory and file
+ * renames. Blob Endpoint does not expose rename API, this class is responsible
+ * for copying the blobs and deleting the source blobs.
+ *
+ * For directory rename, it recursively lists the blobs in the source directory and
+ * copies them to the destination directory.
+ */
+public class BlobRenameHandler extends ListActionTaker {
+
+  public static final Logger LOG = LoggerFactory.getLogger(AbfsClient.class);
+
+  private final String srcEtag;
+
+  private final Path src, dst;
+
+  private final boolean isAtomicRename, isAtomicRenameRecovery;
+
+  private final TracingContext tracingContext;
+
+  private AbfsLease srcAbfsLease;
+
+  private String srcLeaseId;
+
+  private final List<AbfsLease> leases = new ArrayList<>();
+
+  private final AtomicInteger operatedBlobCount = new AtomicInteger(0);
+
+  public BlobRenameHandler(final String src,
+      final String dst,
+      final AbfsClient abfsClient,
+      final String srcEtag,
+      final boolean isAtomicRename,
+      final boolean isAtomicRenameRecovery,
+      final TracingContext tracingContext) {
+    super(new Path(src), abfsClient, abfsClient.getAbfsConfiguration()
+        .getBlobRenameDirConsumptionParallelism(), tracingContext);
+    this.srcEtag = srcEtag;
+    this.tracingContext = tracingContext;
+    this.src = new Path(src);
+    this.dst = new Path(dst);
+    this.isAtomicRename = isAtomicRename;
+    this.isAtomicRenameRecovery = isAtomicRenameRecovery;
+  }
+
+  /**
+   * Orchestrates the rename operation.
+   */
+  public AbfsClientRenameResult execute() throws IOException {
+    PathInformation pathInformation = new PathInformation();
+    final boolean result;
+    if (preCheck(src, dst, pathInformation)) {
+      RenameAtomicity renameAtomicity = null;
+      if (isAtomicRename) {
+        /*
+         * Conditionally get a lease on the source blob to prevent other writers
+         * from changing it. This is used for correctness in HBase when log files
+         * are renamed. When the HBase master renames a log file folder, the lease
+         * locks out other writers.  This prevents a region server that the master
+         * thinks is dead, but is still alive, from committing additional updates.
+         * This is different than when HBase runs on HDFS, where the region server
+         * recovers the lease on a log file, to gain exclusive access to it, before
+         * it splits it.
+         */
+        srcAbfsLease = takeLease(src, srcEtag);
```

Comment:
lease should be released as well in the finally block ?

Reply:
taken.